src/Streamly/Internal/Unicode/Utf8/Type.hs

```diff
@@ -0,0 +1,501 @@
+read :: Unfold m Utf8 Char
+read = undefined
+
+--------------------------------------------------------------------------------
+-- Basic functions
+--------------------------------------------------------------------------------
+
+-- XXX Use K version?
+-- | /O(n)/ Adds a character to the front of a 'Utf8'. This function is more
+-- costly than its 'List' counterpart because it requires copying a new array.
+-- Subject to fusion. Performs replacement on invalid scalar values.
+{-# INLINE cons #-}
+cons :: Char -> Utf8 -> Utf8
+cons c = unstream . Stream.cons c . stream
+
+-- | /O(n)/ Adds a character to the end of a 'Utf8'.  This copies the entire
+-- array in the process, unless fused.  Subject to fusion.  Performs replacement
+-- on invalid scalar values.
+{-# INLINE snoc #-}
+snoc :: Utf8 -> Char -> Utf8
+snoc = undefined
+
+-- XXX NOINLINE because text says so. Need to check why.
+-- XXX copying them to a new Utf8? It might not if space is available in the
+-- first Utf8 (Array).
+-- | /O(n)/ Appends one 'Utf8' to the other by copying both of them into a new
+-- 'Utf8'.  Subject to fusion.
+{-# NOINLINE append #-}
+append :: Utf8 -> Utf8 -> Utf8
+append (Utf8 a) (Utf8 b) = Utf8 $ unsafePerformIO $ Array.spliceExp a b
+
+{-
+{-# RULES
+"TEXT append -> fused" [~1] forall t1 t2.
+    append t1 t2 = unstream (S.append (stream t1) (stream t2))
+"TEXT append -> unfused" [1] forall t1 t2.
+    unstream (S.append (stream t1) (stream t2)) = append t1 t2
+ #-}
+-}
+
+-- | /O(1)/ Returns the first character of a 'Utf8', or 'Nothing' if empty.
+-- Subject to fusion.
+{-# INLINE head #-}
+head :: Utf8 -> Maybe Char
+head = unsafePerformIO . Stream.head . stream
+
+-- XXX INLINE_NORMAL cause text. Need to find the reason.
+-- | /O(1)/ Returns the first character and rest of a 'Utf8', or 'Nothing' if
+-- empty. Subject to fusion.
+{-# INLINE_NORMAL uncons #-}
+uncons :: Utf8 -> Maybe (Char, Utf8)
+uncons = fmap (second unstream) . unsafePerformIO . Stream.uncons . stream
+
+-- | /O(1)/ Returns the last character of a 'Utf8', or 'Nothing' if empty.
+-- Subject to fusion.
+{-# INLINE_NORMAL last #-}
+last :: Utf8 -> Char
+last = undefined
+
+{-
+{-# RULES
+"TEXT last -> fused" [~1] forall t.
+    last t = S.last (stream t)
+"TEXT last -> unfused" [1] forall t.
+    S.last (stream t) = last t
+  #-}
+-}
+
+-- | /O(1)/ Returns all characters after the head of a 'Utf8', or 'Nothing' if
+-- empty.  Subject to fusion.
+{-# INLINE_NORMAL tail #-}
+tail :: Utf8 -> Maybe Utf8
+tail = fmap unstream . unsafePerformIO . Stream.tail . stream
+
+{-
+{-# RULES
+"TEXT tail -> fused" [~1] forall t.
+    tail t = unstream (S.tail (stream t))
+"TEXT tail -> unfused" [1] forall t.
+    unstream (S.tail (stream t)) = tail t
+ #-}
+-}
+
+-- | /O(1)/ Returns all but the last character of a 'Utf8', or 'Nothing' if
+-- empty.  Subject to fusion.
+{-# INLINE_NORMAL init #-}
+init :: Utf8 -> Maybe Utf8
+init = fmap unstream . unsafePerformIO . Stream.init . stream
```

Comment:
If we can write a routine to decode utf8 in reverse then we can just decode the last char from the end of the array and then slice it.

Otherwise, use `last` on the stream, get the encoded length of the last char and use that to slice it.